File: Week 5/1. Service/ros2_ws/src/pkg_py/pkg_py/add_two_ints_oop_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from functools import partial

from example_interfaces.srv import AddTwoInts

class AddTwoIntegersClient(Node):
    def __init__(self):
        super().__init__("add_two_ints_oop_clients")
        self.call_add_two_ints_server(6, 7)

    def call_add_two_ints_server(self, a, b):
        client = self.create_client(AddTwoInts, "add_two_ints")
        while not client.wait_for_service(1.0):
            self.get_logger().warn("Waiting for Server...")

        request = AddTwoInts.Request()
        request.a = a
        request.b = b

        future = client.call_async(request)
        future.add_done_callback(partial(self.call_add_two_ints_callback,a=a,b=b))

    def call_add_two_ints_callback(self, future, a, b):
        try:
            response = future.result()
            self.get_logger().info(f'{a} + {b} = {response.sum}')
        except Exception as e:
            self.get_logger().error("Service call failed: %r",(e,))


def main(args=None):
    rclpy.init(args=args)
    node = AddTwoIntegersClient()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: Week 5/1. Service/ros2_ws/src/pkg_cpp_py/src/add_two_ints_oop_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"

using namespace std::chrono_literals;

class AddTwoIntsOOP : public rclcpp::Node
{
public:
    AddTwoIntsOOP(int arg1, int arg2) : Node("add_two_ints_oop_client")
    {
        RCLCPP_INFO(this->get_logger(), "Client ready to send request..");

        thread = std::thread(std::bind(&AddTwoIntsOOP::callAddTwoIntsService, this, arg1, arg2));
    }

    void callAddTwoIntsService(int a, int b)
    {
        client = this->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");

        while (!client->wait_for_service(1s))
        {
            RCLCPP_WARN(this->get_logger(), "Service not available, waiting again...");
        }

        auto request = example_interfaces::srv::AddTwoInts::Request::SharedPtr();
        request->a = a;
        request->b = b;

        auto future = client->async_send_request(request);

        try
        {
            auto response = future.get();
            RCLCPP_INFO(this->get_logger(), "The sum of %d and %d is %d", a,b,response->sum);
        }
        catch (const std::exception &e)
        {
            RCLCPP_ERROR(this->get_logger(), "Oops! Service call failed.");
        }
    }

private:
    rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedPtr client;
    std::thread thread;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<AddTwoIntsOOP>(atoll(argv[1]),atoll(argv[2]));
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
